onnxruntime/core/common/threadpool.cc

```diff
@@ -1,37 +1,118 @@
-// Copyright (c) Microsoft Corporation. All rights reserved.
-// Licensed under the MIT License.
 
-#include "core/platform/threadpool.h"
-#include "core/common/common.h"
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
 
-#include <cassert>
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
 
-#if defined(__GNUC__)
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wunused-parameter"
-#else
-#pragma warning(push)
-#pragma warning(disable : 4267)
+#include <memory>
+
+#include "core/platform/threadpool.h"
+#include "core/common/common.h"
+#ifndef EIGEN_USE_THREADS
+#define EIGEN_USE_THREADS
 #endif
-#include <unsupported/Eigen/CXX11/src/ThreadPool/Barrier.h>
-#if defined(__GNUC__)
-#pragma GCC diagnostic pop
-#else
-#pragma warning(pop)
+#include "core/util/eigen_common_wrapper.h"
+#include "core/platform/EigenNonBlockingThreadPool.h"
+#include "core/platform/ort_mutex.h"
+
+namespace onnxruntime {
+namespace {
+class BlockingCounter {
+ public:
+  BlockingCounter(int initial_count) : state_(initial_count << 1), notified_(false) {
+    ORT_ENFORCE(initial_count >= 0);
+#ifndef NDEBUG
+    ORT_ENFORCE(((initial_count << 1) >> 1) == initial_count);
 #endif
+  }
 
-using Eigen::Barrier;
+  ~BlockingCounter() = default;
 
-namespace onnxruntime {
+  inline void DecrementCount() {
+    unsigned int v = state_.fetch_sub(2, std::memory_order_acq_rel) - 2;
+    if (v != 1) {
+#ifndef NDEBUG
+      ORT_ENFORCE(((v + 2) & ~1) != 0);
+#endif
+      return;  // either count has not dropped to 0, or waiter is not waiting
+    }
+    std::lock_guard<OrtMutex> l(mu_);
+    notified_ = true;
```

Comment:
> notified_ = true; [](start = 3, length = 18)

Is this a one time use counter? It better be.

Reply:
Yes, it is.

---
In reply to: [391271213](https://github.com/microsoft/onnxruntime/pull/3153#discussion_r391271213) [](ancestors = 391271213)